src/ore/src/retry.rs

```diff
@@ -219,6 +221,163 @@ impl Default for Retry {
+                } else if elapsed + this.next_backoff.unwrap() > max_duration {
+                    *this.next_backoff = Some(max_duration - elapsed);
+                }
+            }
+            _ => (),
+        }
+
+        let state = RetryState {
+            i: *this.i,
+            next_backoff: *this.next_backoff,
+        };
+        if let Some(d) = *this.next_backoff {
+            this.sleep.reset(tokio::time::Instant::now() + d);
+        }
+        *this.i += 1;
+        Poll::Ready(Some(state))
+    }
+}
+
+/// Wrapper of a `Reader` factory that will automatically retry and resume reading an underlying
+/// resource in the events of errors according to a retry schedule.
+#[pin_project]
+#[derive(Debug)]
+pub struct RetryReader<F, U, R> {
+    factory: F,
+    offset: usize,
+    error: Option<std::io::Error>,
+    #[pin]
+    retry: Retry,
+    #[pin]
+    state: RetryReaderState<U, R>,
+}
+
+#[pin_project(project = RetryReaderStateProj)]
+#[derive(Debug)]
+enum RetryReaderState<U, R> {
+    Waiting,
+    Creating(#[pin] U),
+    Reading(#[pin] R),
+}
+
+impl<F, U, R> RetryReader<F, U, R>
+where
+    F: FnMut(RetryState, usize) -> U,
+    U: Future<Output = Result<R, std::io::Error>>,
+    R: AsyncRead,
+{
+    /// Uses the provided `Reader` factory to construct a `RetryReader` with the default `Retry`
+    /// settings.
+    ///
+    /// The factory will be called once at the beginning and subsequently every time a retry
+    /// attempt is made. The factory will be called with a single `usize` argument representing the
+    /// offset at which the returned `Reader` should resume reading from.
+    pub fn new(factory: F) -> Self {
+        Self::with_retry(factory, Retry::default())
+    }
+
+    /// Uses the provided `Reader` factory to construct a `RetryReader` with the passed `Retry`
+    /// settings. See the documentation of [RetryReader::new] for more detais.
+    pub fn with_retry(factory: F, retry: Retry) -> Self {
+        Self {
+            factory,
+            offset: 0,
+            error: None,
+            retry,
+            state: RetryReaderState::Waiting,
+        }
+    }
+}
+
+impl<F, U, R> AsyncRead for RetryReader<F, U, R>
+where
+    F: FnMut(RetryState, usize) -> U,
+    U: Future<Output = Result<R, std::io::Error>>,
+    R: AsyncRead,
+{
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        loop {
+            let mut this = self.as_mut().project();
+            use RetryReaderState::*;
+            match this.state.as_mut().project() {
+                RetryReaderStateProj::Waiting => match ready!(this.retry.as_mut().poll_next(cx)) {
+                    None => {
+                        return Poll::Ready(Err(this
+                            .error
+                            .take()
+                            .expect("retry produces at least one element")))
```

Comment:
The reason this is safe is because we should have made it through `Creating` or `Reading` at least once before there is an empty result?

Reply:
Yes, exactly. The retry stream guarantees that there will at least one try